### PR TITLE
docs(process): prose-discipline rules, an upstream citation registry, and a checker

### PR DESCRIPTION
**Pull request by dknauss**
#317 needed three rounds of corrections. Every one was to prose, none to code — and two of the three were caught by Codex rather than by tests, a review agent, or my own sweeps.

## The pattern

Each false claim was a **true narrow observation inflated by a scope word**.

| Verified | What shipped |
|---|---|
| L280 has a conditional icon on this button | "core swaps `starFilled/starEmpty` **on this very toggle**" — L326 is a different control |
| Pinned buttons get no semantic background | "**the sole colour** in that header is Publish" — the Inserter is the same blue |
| The CSS painted one state | "**the only** painted state" — true for four hours |

The citation lends its credibility to the whole sentence, including the part nothing checked. That is what makes this hard to catch while writing.

## What this adds

**Rules** — `AGENTS.md` → "Prose Discipline", with a summary and pointer in `CLAUDE.md`. Six rules, each naming the incident that produced it so none of it reads as generic hygiene. The two with the most leverage: cite the **enclosing symbol** (you cannot fill that field from a grep hit — that is the mechanism), and treat **scope words as claims** needing evidence.

**Registry** — `docs/upstream-sources.md`, the same arrangement `current-metrics.md` gives counts. One row per third-party claim; prose carries a summary and an ID instead of a seventh copy of the URL. `GB-PIN-STAR` records the L326 misattribution as a **counter-example** so it cannot recur. `admin/css/wp-sudo-editor-indicator.css` is migrated as the worked example.

**Checker** — `bin/verify-sources.sh` / `composer verify:sources` + a nightly workflow. **Fails** when a snippet has vanished upstream; **warns** when it has only moved, because trunk drifts constantly and a noisy check gets switched off. Not a PR gate (it fetches over the network, and a required check that fails for unrelated reasons is one people learn to force past) and not `continue-on-error` (this repo already has `e2e-visual.yml` reporting "success" over four failing baselines for weeks).

## Verified, not asserted

The first run found **two bugs in my own parser** — prose lines containing pipes parsed as table rows, and the `\|` escape breaking field splitting. After fixing, it found **two genuine line drifts** in citations already on `main`, which `--fix-lines` corrected. I also tested the negative path: a deliberately false snippet and a row missing its enclosing symbol both fail, exit 1. Current state: 9 citations checked, zero warnings.

## Second commit: a stale POT, and a check nobody ran

`bin/verify-i18n.sh` **fails on `main` right now** — `wp-sudo-session-indicator.js:202` should be `:205`. I hand-patched five POT line references across #317 as comment blocks shifted the strings, and missed a sixth. `composer i18n:make-pot` exists and would have been faster and correct; that is now a rule in this branch, written before I discovered I had broken this.

CI never caught it because **`composer verify:i18n` was wired into no workflow at all**. Now runs beside `verify:metrics`. A check nobody runs is indistinguishable from a check that does not exist — the second instance of that in two days.

## One thing deliberately left out

`CLAUDE.md` and `AGENTS.md` both carry a "Verification Requirements" section and **they have already diverged**: 74 lines vs 63. `CLAUDE.md` has an Internal-architectural-counts subsection that `AGENTS.md` compresses, and a release step to bump the `blueprint.json` Playground target that `AGENTS.md` **lacks entirely** — so an agent following `AGENTS.md` at release time would ship a stale Playground demo. Codex cites `AGENTS.md`; Claude Code loads `CLAUDE.md`.

That is a release-correctness bug with different urgency and reviewers, so it is not folded in here. Worth its own issue.

Refs #288, #317

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01V4kcGFoqkxivzk9BSRUhEs

### CODE REVIEW

File: bin/verify-sources.sh [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+orphans="$(grep -rIl --exclude-dir={vendor,node_modules,.git,vendor_test,.tmp,.planning} \
+	-e 'raw\.githubusercontent\.com/WordPress' "$REPO_ROOT" 2>/dev/null \
+	| grep -v 'docs/upstream-sources.md' \
+	| grep -v 'docs/llm-lies-log.md' \
+	| grep -v 'bin/verify-sources.sh' \
+	| grep -vE '/(AGENTS|CLAUDE)\.md$' || true)"
+if [ -n "$orphans" ]; then
+	while IFS= read -r f; do
+		[ -n "$f" ] || continue
+		add_warning "citation outside the registry: ${f#"$REPO_ROOT/"}"
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh]: Applied in 7a9c633. A new unregistered upstream citation is now a hard failure, so the scheduled job goes red when prose bypasses the registry. The 8 pre-existing files are grandfathered to warnings (GRANDFATHERED_ORPHANS) so the mandatory-registration rule was adopted without red-building the migration backlog; the list shrinks as files migrate. Covered by the self-link regression cases in tests/verify-sources/run.sh.

File: bin/verify-sources.sh [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+fi
+
+for w in "${warnings[@]:-}"; do
+	[ -n "$w" ] && echo "WARN: $w"
+done
+
+if [ ${#failures[@]} -gt 0 ] && [ -n "${failures[0]:-}" ]; then
+	if [ "$network_ok" = "0" ] && [ "$OFFLINE_OK" = "1" ]; then
+		echo "SKIPPED: upstream unreachable and --offline-ok was passed."
+		exit 0
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh]: Addressed in 7a9c633. Content failures were already preserved under --offline-ok via the hard/network split (only hard_failures==0 skips). The other half — an HTTP 404 misclassified as unreachable — is fixed by branching on %{http_code} before the curl exit code (only status 000/empty is network now). Regressions: 'content failure preserved under --offline-ok' and 'http 404 with curl error is not suppressed by --offline-ok'.

File: .github/workflows/phpunit.yml [hunk 3]
```diff
@@ -308,6 +308,12 @@ jobs:
       - name: Verify metrics file drift
         run: composer verify:metrics
 
+      # The script existed but ran nowhere, so a stale POT reached main in #317:
+      # line references had been hand-patched instead of regenerated, and one was
+      # missed. Cheap to run, and it only fails when the POT is genuinely out of date.
+      - name: Verify POT file drift
+        run: composer verify:i18n
```
[the remark this answers is not in the document]
Comment (dknauss) [reply .github/workflows/phpunit.yml:344]: Already satisfied: the code-quality job installs WP-CLI (pinned 2.12.0 + sha256 verify) in the step immediately before 'Verify POT file drift' — added in b07d2a9, still present in .github/workflows/phpunit.yml. verify:i18n therefore has wp available. No change needed.

File: bin/verify-sources.sh [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+	set +e
+	status="$(curl -sSL --max-time 30 -o "$path" -w '%{http_code}' "$url" 2>/dev/null)"
+	code=$?
+	set -e
+
+	if [ "$code" -ne 0 ]; then
+		# Transport failure: DNS, connect, TLS, timeout, proxy, reset. No answer at
+		# all, so nothing can be concluded about the claim.
+		rm -f "$path"
+		FETCH_ERR="network"
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh]: Fixed in 7a9c633: fetch() now branches on %{http_code} first and treats only status 000/empty as a network outage. A 404 that also yields a nonzero curl exit (truncated/reset body -> exit 18/56 with http_code 404) is filed as an HTTP finding that --offline-ok will not swallow. Regression 'http 404 with curl error still fails as http'.

File: bin/verify-sources.sh [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+# matters.
+# Results come back in FETCH_PATH / FETCH_ERR rather than on stdout: `$(fetch ...)`
+# runs the function in a SUBSHELL, so anything it assigns is discarded when that
+# subshell exits. An earlier revision returned the path on stdout and set FETCH_ERR as
+# a side effect, which meant the caller always saw an empty FETCH_ERR and classified
+# every failure as an HTTP error — including a plain unreachable network.
+FETCH_ERR=""
+FETCH_PATH=""
+FETCH_STATUS=""
+fetch() {
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh:87]: Added in 7a9c633: tests/verify-sources/run.sh — an offline harness (fake curl on PATH, script symlinked so REPO_ROOT resolves inside a throwaway sandbox), 34 assertions covering duplicate IDs, ambiguous match, incomplete rows, HTTP 404 (clean and nonzero-exit) vs network, offline content failures, dangling id, non-GB id rejection, self-link filtering, and non-raw hosts. Wired as composer test:sources in the code-quality gate and before the nightly live check.

File: bin/verify-sources.sh [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+	# instruction about escaping them, for one), so leading `|` is the row test.
+	case "$raw_row" in '|'*) ;; *) continue ;; esac
+
+	row="${raw_row//\\|/$ESC}"
+	IFS='|' read -r _ id url line snippet symbol claim _rest <<< "$row"
+
+	id="$(clean_field "${id:-}")"
+	[ -n "$id" ] || continue
+	[ "$id" != "ID" ] || continue
+	case "$id" in ---*|:--*) continue ;; esac
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh:182]: Fixed in 7a9c633: a duplicate ID now fails immediately (space-delimited seen-ids set, bash-3.2 safe), before fetch and before --fix-lines can act. Regression 'duplicate IDs fail'.

File: bin/verify-sources.sh [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+		else
+			add_failure "$id: $url returned HTTP $FETCH_STATUS — the file has moved or been deleted upstream
+        claim now unsupported: $claim"
+		fi
+		continue
+	fi
+
+	file_path="$FETCH_PATH"
+	needle="$(printf '%s' "$snippet" | squeeze)"
+	hit="$(squeeze < "$file_path" | grep -nF -- "$needle" | head -n1 || true)"
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh]: Fixed in 7a9c633: more than one upstream match is now a hard failure asking for a fragment unique to the recorded symbol, instead of head -n1 silently picking the first copy (which --fix-lines could then repoint). Regression 'ambiguous match fails'.

File: bin/verify-sources.sh [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+# A first cut matched per file and warned about 14, of which 6 were noise — including
+# verify-sources.yml, i.e. the tool warning about itself. A check whose own comment
+# says noisy checks get switched off does not get to run at a 43% false-positive rate.
+orphans="$(grep -rInIE \
+	-e 'raw\.githubusercontent\.com' \
+	-e 'plugins\.svn\.wordpress\.org' \
+	-e 'github\.com/[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+/blob/' \
+	--exclude-dir={vendor,node_modules,.git,vendor_test,.tmp,.planning} \
+	"$REPO_ROOT" 2>/dev/null \
+	| grep -viE 'dknauss(/|%2f)sudo' \
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh:385]: Fixed in 7a9c633: the orphan scan uses grep -o with owner-capturing URL patterns, so the self-link filter runs per matched URL rather than per line. A comparison row citing both this repo and a third-party raw URL keeps the third-party orphan. Regressions 'self-link does not drop co-located third-party citation' and 'pure self-link is not an orphan'.

File: bin/verify-sources.sh [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+# --- dangling registry IDs -------------------------------------------------------
+# The registry only helps if the IDs prose points at exist. Replacing URLs with IDs
+# creates this failure mode, so it ships with the check for it: a renamed or deleted
+# row would otherwise leave a comment pointing at nothing, silently.
+# -I to match the orphan scan below: a binary containing these bytes would otherwise
+# be a hard failure. No --exclude for this script: an example ID written into these
+# comments must be caught too — the registry's how-to section once pointed at an ID
+# that did not exist, and an example here would have been just as invisible. Which is
+# why this comment names no ID.
+referenced="$(grep -rhoIE '\bGB-[A-Z0-9][A-Z0-9-]+' \
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh:327]: Fixed in 7a9c633 by enforcing one universal ID syntax: every registry ID must use the GB- prefix, and a row that would mint an ID outside it is rejected. A dangling reference therefore cannot exist outside the scanned prefix. Documented in docs/upstream-sources.md; regression 'non-GB registry id rejected'.

File: docs/upstream-sources.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+4. In prose, write `see GB-ICON-SWAP in docs/upstream-sources.md` rather than another
+   copy of the URL and line.
+
+Escape any `|` inside a snippet as `\|`.
+
+## Registry
+
+| ID | Raw URL | Line | Must contain | Enclosing symbol | Claim it supports |
+|---|---|---|---|---|---|
+| GB-PRESSED-FILL | https://raw.githubusercontent.com/WordPress/gutenberg/trunk/packages/components/src/button/style.scss | 347 | `background: $components-color-foreground;` | `&.is-pressed` block in `.components-button` | Core applies no *semantic* background colour to a pinned-item button — the pressed state fills with a neutral foreground token. (Not the only background core sets: `:disabled`/`[aria-disabled]` gets `$gray-600`, and `.is-tertiary:hover` gets an accent mix in the icon-labels mode. None of them encode meaning, which is the point.) This is why WP Sudo carries state on the glyph and not a chip. |
```
[the remark this answers is not in the document]
Comment (dknauss) [reply docs/upstream-sources.md]: Fixed in 7a9c633: the GB-PRESSED-FILL claim is narrowed to what the single declaration proves — the pressed state fills with a neutral foreground token, not a status colour. The un-falsifiable 'core applies no semantic background anywhere' is reframed as the design reading built on that line, not a verified upstream absence.

File: docs/upstream-sources.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+| GB-SCOPED-SUBSCRIBE | https://raw.githubusercontent.com/WordPress/gutenberg/trunk/packages/data/src/registry.ts | 62 | `const subscribe = (` | `createRegistry` | `subscribe( listener, storeNameOrDescriptor )` accepts a store scope, so the indicator's `core/preferences` listener does not run on every editor keystroke. |
+| GB-AREA-IDENTIFIER | https://raw.githubusercontent.com/WordPress/gutenberg/trunk/packages/interface/src/components/complementary-area-toggle/index.js | 45 | `const identifier = identifierProp \|\| ` | `ComplementaryAreaToggle` | `aria-controls` is `{plugin-name}:{sidebar-name}`, which is what makes the CSS selector stable and locale-independent. |
```
[the remark this answers is not in the document]
Comment (dknauss) [reply docs/upstream-sources.md]: Fixed in 7a9c633: GB-SCOPED-SUBSCRIBE now cites L64 'storeNameOrDescriptor?: StoreNameOrDescriptor' (subscribe's optional scope parameter) and GB-AREA-IDENTIFIER now cites '${ context.name }/${ name }' (the identifier fallback). Both verified live at their recorded lines via composer verify:sources.

File: bin/verify-sources.sh [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+	# fails as "no enclosing symbol recorded", sending the reader after the wrong bug.
+	if [ -n "$(clean_field "${_rest:-}")" ]; then
+		add_failure "$id: row has more columns than the table defines — escape any literal | inside a snippet as \\|"
+		continue
+	fi
+
+	if [ -z "$url" ] || [ -z "$snippet" ]; then
+		add_failure "$id: row is missing a URL or a snippet"
+		continue
+	fi
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh]: Fixed in 7a9c633: a row must carry a URL, a Line, a snippet, and a Claim (the enclosing symbol is still required separately after the URL is validated). Regressions 'missing Line fails' and 'missing claim fails'.

File: bin/verify-sources.sh [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+# The registry only helps if the IDs prose points at exist. Replacing URLs with IDs
+# creates this failure mode, so it ships with the check for it: a renamed or deleted
+# row would otherwise leave a comment pointing at nothing, silently.
+# -I to match the orphan scan below: a binary containing these bytes would otherwise
+# be a hard failure. No --exclude for this script: an example ID written into these
+# comments must be caught too — the registry's how-to section once pointed at an ID
+# that did not exist, and an example here would have been just as invisible. Which is
+# why this comment names no ID.
+referenced="$(grep -rhoIE '\bGB-[A-Z0-9][A-Z0-9-]+' \
+	--exclude-dir={vendor,node_modules,.git,vendor_test,.tmp} \
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh]: Fixed in 7a9c633: .claude is excluded from both the repo-wide GB-id scan and the orphan scan (it holds sibling worktree checkouts). The checker's own test dir is excluded the same way, since its fixtures necessarily contain example GB-* IDs and raw URLs.

File: CHANGELOG.md [hunk 14]
```diff
@@ -2,7 +2,7 @@
...
...time in; and because two of the three states carry no colour at all, the states stay distinguishable in greyscale, under colour-vision deficiency, and in forced-colors mode by construction rather than by supplement. One mode is handled separately: core's **Show button text labels** preference replaces this button's icon with its own `check` glyph for every state (`icon={ showIconLabels ? check : icon }`, L280) and suppresses the tooltip with it (`showTooltip={ ! showIconLabels }`, L281), which would leave that cohort with neither a shape channel nor a hover affordance — active and inactive reduced to one appearance, the very bug this feature exists to fix. There, and only there, the green active chip returns: colour is the only channel core leaves. The default view keeps no semantic colour but the urgent red. Pinned by INDICATOR-09, which also asserts the chip cannot leak into the default view. The trade-off is deliberate and worth naming: a 20 px glyph change is less noticeable in peripheral vision than a green-to-red chip, so this trades some at-a-glance salience for fitting the surface it lives on. Pinned by INDICATOR-06/07/08/09 and three visual baselines, one per state. (#288)
```
[the remark this answers is not in the document]
Comment (dknauss) [reply CHANGELOG.md]: Fixed in 7a9c633: the Unreleased #288 entry now references GB-PRESSED-FILL / GB-ICON-SWAP / GB-SELECTED-ICON / GB-NO-TOOLTIP instead of embedding duplicate file/line/snippet citations.

File: docs/upstream-sources.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+| GB-PIN-STAR | https://raw.githubusercontent.com/WordPress/gutenberg/trunk/packages/interface/src/components/complementary-area/index.js | 326 | `icon={ isPinned ? starFilled : starEmpty }` | `Button.interface-complementary-area__pin-unpin-item`, in the panel header | **Counter-example, recorded so it is not miscited again.** This is the pin/unpin star inside the complementary-area panel header — a different control from the pinned toolbar button. #317's first commit cited it as precedent for the toolbar toggle; that was wrong, and was corrected on the same branch before merge. |
+| GB-SCOPED-SUBSCRIBE | https://raw.githubusercontent.com/WordPress/gutenberg/trunk/packages/data/src/registry.ts | 62 | `const subscribe = (` | `createRegistry` | `subscribe( listener, storeNameOrDescriptor )` accepts a store scope, so the indicator's `core/preferences` listener does not run on every editor keystroke. |
```
[the remark this answers is not in the document]
Comment (dknauss) [reply docs/upstream-sources.md]: Done: 7a9c633's commit message records the live raw.githubusercontent.com sources for the changed registry rows (registry.ts L64, complementary-area-toggle L45, button/style.scss L347) and notes that all 9 rows re-verified live via composer verify:sources with no drift.

File: docs/upstream-sources.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+Every claim this repo makes about **third-party code** lives here, once. Code comments,
+changelog entries, specs, and design docs carry a one-line summary and point at an ID in
+this table — they do not restate the citation. This is the same arrangement
+`docs/current-metrics.md` uses for counts, and for the same reason: a fact duplicated
+across seven files is a fact that will be wrong in six of them within a month.
+
+`bin/verify-sources.sh` checks every row against the live upstream file. It fails when a
+snippet has **disappeared** (the claim is now false) and warns when a snippet has merely
+**moved** (the line number drifted; upstream trunk moves constantly and that alone is not
```
[the remark this answers is not in the document]
Comment (dknauss) [reply docs/upstream-sources.md]: Fixed in 7a9c633: docs/upstream-sources.md now describes a missing snippet as 'unsupported / needs re-verification' (upstream can reformat a cited expression across lines, dropping the one-line snippet without changing behaviour), matching the script's own failure output rather than calling the claim 'false'.

File: bin/verify-sources.sh [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+		FETCH_ERR="network"
+		return 1
+	fi
+
+	case "$status" in
+		2??)
+			FETCH_PATH="$path"
+			return 0
+			;;
+		*)
```
[the remark this answers is not in the document]
Comment (dknauss) [reply bin/verify-sources.sh:151]: Fixed in 3b6e666: fetch() now separates permanent drift (404/410, hard failure) from a retryable availability failure (429/5xx/403 rate limit), which is counted with the network failures so --offline-ok can suppress a temporary outage and is reported as 'temporarily unavailable, not treated as drift' rather than 'moved or deleted'. Unreachable (000/empty) stays its own case. curl still omits --fail so %{http_code} is authoritative. Without --offline-ok the run still fails on a transient outage. Regressions: 'transient 5xx is not reported as deleted', '5xx/429 suppressed by --offline-ok', '410 Gone is permanent drift'.
